### PR TITLE
Determine incomplete/invalidated periods for evolution graphs

### PR DESCRIPTION
**Pull request by mneudert**
### Description:

Provides archive state information for evolution graphs to display information about incomplete or invalidated data points.

#### State Selection

An `Archive\DataCollection` contains information about "invalidated" or "incomplete" data based on several checks:

- `invalidated`
  - any of the `done` archive entries was `invalidated`
- `incomplete`
  - `ts_archived` has a date earlier than the end of the period
  - a `report_to_invalidate` entry is found for any day of the period (requests have been tracked after the archive was written)

State is only provided for periods with data (matching the availability of `ts_archived` information), and all periods containing a `report_to_invalidate`.

#### State Retrieval

DataTables retrieved using the `Archive::getDataTable*` functions will contain state information.

Unless `fetch_archive_state=1` is passed in the request, this state will be automatically removed by the `DataTablePostProcessor`, making it opt-in.

#### Evolution Graph Display

The state informationen is passed to evolution graphs for display (see #21694).

If no state information is available (no archive found), the displayed values will be flagged as "complete", unless the point follows a point marked as "incomplete". In the latter case, the "incomplete" flag will be used for the missing archive.

Refs DEV-17568

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

### CODE REVIEW

File: core/Archive.php [hunk 1]
```diff
@@ -572,35 +601,104 @@ protected function get($archiveNames, $archiveDataType, $idSubtable = null)
...
+                        continue;
+                    }
+
+                    $archiveState = $archiveStates[$idSite][$period][$archiveId];
+
+                    // pick "worst" state
+                    switch ($archiveState) {
+                        case ArchiveWriter::DONE_OK:
+                            if (null === $metaArchiveState) {
+                                $metaArchiveState = DataTable::ID_ARCHIVE_STATE_COMPLETE;
```
Comment (caddoo) [inline core/Archive.php]: There is a lot of nesting, do you think it makes sense to extract parts of this to private methods with sensible names?

**Review by bx80 — DISMISSED** [review]
Great work @mneudert :+1: Manual testing was fine, the new automated tests are passing and I can't see any issues with the code.

Could an archive be considered both incomplete and partial? If so, which state would take precedence and do we have tests to cover this?
When reverting [PR 14639](https://github.com/matomo-org/matomo/pull/14639) do we need to change anything in custom reports?

It's a fairly impactful change so I think it would be good to get another review to be sure we've covered all the bases.

File: core/API/DataTablePostProcessor.php [hunk 2]
```diff
@@ -146,6 +147,23 @@ private function convertSegmentValueToSegment(DataTableInterface $dataTable)
         return $dataTable;
     }
 
+    /**
+     * @param DataTableInterface $dataTable
+     * @return DataTableInterface
+     */
+    public function applyArchiveStateFilter($dataTable)
```
Comment (kniziol) [inline core/API/DataTablePostProcessor.php]:
```suggestion
    public function applyArchiveStateFilter(DataTableInterface $dataTable): DataTableInterface
```
Will it work?

File: core/Archive.php [hunk 3]
```diff
@@ -572,35 +601,104 @@ protected function get($archiveNames, $archiveDataType, $idSubtable = null)
...
+                $periodStr,
+                $row['name'],
+                $row['value'],
+                [
+                    DataTable::ARCHIVED_DATE_METADATA_NAME => $row['ts_archived'],
+                ]
+            );
+        }
+
+        foreach ($addMetaForPeriods as $idSite => $metaPeriods) {
```
Comment (kniziol) [inline core/Archive.php]: It might be worth moving the loop to a separate, private method if possible

File: core/Archive.php [hunk 4]
```diff
@@ -572,35 +601,104 @@ protected function get($archiveNames, $archiveDataType, $idSubtable = null)
...
+
+                foreach ($archiveIds[$period] as $archiveId) {
+                    if (!isset($archiveStates[$idSite][$period][$archiveId])) {
+                        continue;
+                    }
+
+                    $archiveState = $archiveStates[$idSite][$period][$archiveId];
+
+                    // pick "worst" state
+                    switch ($archiveState) {
```
Comment (kniziol) [inline core/Archive.php]: The `switch ($archiveState) {` block of code is a good candidate for a new method:
```php
private function metaArchiveStateFromArchiveState(string $archiveState): string
{
    switch ($archiveState) {
        // ...
    }
}
```
The logic of the `switch` can also be moved into a separate, new class. It'd be more accessible to cover by tests.

File: core/Archive.php [hunk 5]
```diff
@@ -572,35 +601,104 @@ protected function get($archiveNames, $archiveDataType, $idSubtable = null)
...
      *
      * @param string[] $archiveNames
-     * @return array
+     *
+     * @return array An array with two arrays:
+     *               - archive ids
+     *               - archive doneFlag values
      */
-    private function getArchiveIds($archiveNames)
+    private function getArchiveIdsAndStates($archiveNames): array
```
Comment (kniziol) [inline core/Archive.php]:
```suggestion
    private function getArchiveIdsAndStates(array $archiveNames): array
```

File: tests/PHPUnit/Fixtures/RawArchiveDataWithTempAndInvalidated.php [hunk 6]
```diff
@@ -501,26 +501,27 @@ private function setDateMonthAndYear($dateString, Date $archiveDate)
         return $archiveDate->toString('Y-m') . '-' . Date::factory($dateString)->toString('d');
     }
 
-    public function assertTemporaryArchivesPurged($isBrowserTriggeredArchivingEnabled, Date $date)
+    public function assertErroredArchivesPurged(Date $date)
```
Comment (kniziol) [inline tests/PHPUnit/Fixtures/RawArchiveDataWithTempAndInvalidated.php]:
```suggestion
    public function assertErroredArchivesPurged(Date $date): void
```

File: tests/PHPUnit/Fixtures/RawArchiveDataWithTempAndInvalidated.php [hunk 7]
```diff
@@ -501,26 +501,27 @@ private function setDateMonthAndYear($dateString, Date $archiveDate)
...
-            $expectedPurgedArchives = array(1,2,3,4,6,7,10); // only archives from 2 hours before "now" are purged
-        } else {
-            $expectedPurgedArchives = array(1,2,3,4,7); // only archives before start of "yesterday" are purged
-        }
-
+        $expectedPurgedArchives = array(1,2,3,4);
         $this->assertArchivesDoNotExist($expectedPurgedArchives, $date);
     }
 
+    public function assertTemporaryArchivesPurged(Date $date)
```
Comment (kniziol) [inline tests/PHPUnit/Fixtures/RawArchiveDataWithTempAndInvalidated.php]:
```suggestion
    public function assertTemporaryArchivesPurged(Date $date): void
```

File: core/API/DataTablePostProcessor.php [hunk 8]
```diff
@@ -146,6 +147,23 @@ private function convertSegmentValueToSegment(DataTableInterface $dataTable)
...
 
+    /**
+     * @param DataTableInterface $dataTable
+     * @return DataTableInterface
+     */
+    public function applyArchiveStateFilter(DataTableInterface $dataTable): DataTableInterface
+    {
+        $fetchArchiveState = Common::getRequestVar('fetch_archive_state', '0', 'string', $this->request);
+
+        if ('0' === $fetchArchiveState) {
```
Comment (sgiehl) [inline core/API/DataTablePostProcessor.php]: I guess this parameter is meant to be some sort of boolish (0 / 1). We should maybe directly use the newer Request class here:
```suggestion
        $fetchArchiveState = (new \Piwik\Request($this->request))->getBoolParameter('fetch_archive_state', false);

        if (false === $fetchArchiveState) {
```

**Review by sgiehl — COMMENTED** [review]
Only had a rough look through the code, as there were reviews done before. Overall looks good to me, only left one minor comment for improvement.

File: core/Archive.php (hunk 1 above)
Comment (mneudert) [reply core/Archive.php]: Did a bunch of cleanup. Now there is a separate class for the archive state constants and two private methods to populate the collection (one for data, one for state metadata).

**Comment by mneudert** [comment]
Added handling to ignore partial plugin archives, they should auto invalidate themselves, only taking "full" archives into account for displaying.

Tests for CustomReports did not show any change for me locally, so I think we are safe in that regard.

**Review by sgiehl — COMMENTED** [review]
To me the changes look legit and should solve the requirements to exactly know when an archive is fully complete or just temporary.

@diosmosis @tsteur As you both have the most experience with the archiving process it would be awesome and very helpful if you could at least do a quick sense check if the changes here make sense, or if you see any possible problems with reintroducing the `temporary` archive state. I actually can't remember why exactly that state had been removed before.

**Comment by diosmosis** [comment]
If I remember correctly, temporary archives were just used for dates that included today. I think they were deemed redundant.

**Comment by tsteur** [comment]
Indeed, temporary archives were only for "today/this week/this month/this year" as for any historical date whenever we archive it would be "finished" automatically as you can't know if there will be another request coming or not. I'd say as well they were likely redundant and unsure if they would have caused any issues with partial archives.

I haven't looked at the code. Got a question though as temporary archives are basically for "today". I can see that it might help for "yesterday" to show it correctly in a graph but unsure if that works for other days. If we're reintroducing temporary archives, and someone tracks data for 3 days ago making the report for 3 days ago invalid, will this be correctly show as "incomplete" in the UI?

Maybe something to think about is if it's possible to eg rely on `archive_invalidations` table and eg `getRememberedArchivedReportsThatShouldBeInvalidated()` to detect if an archive is incomplete or not. This could simplify it maybe. Didn't think too much about it though if there are any edge cases to consider.

**Comment by mneudert** [comment]
With the change as is, an archive containing today would be internally stored as temporary, instead of complete. They are not as fully temporary as before, though. If a request comes in for three days ago, it will NOT update the archive state from complete to temporary. Data being tracked for past days will only be available through the `getRememberedArchivedReportsThatShouldBeInvalidated()` method you mentioned, and not modify the archive between archiving runs.

If we should to take these requests into account, it is probably the most simple solution to switch to a plain `(ts_archived < period end) or (getRememberedArchivedReports...)` detection. Essentially trading an archiving change with an additional DB query to the option table, moving the implementation risk from an archiving change to the performance impact of a single LIKE-query.

So, two assumptions for simplification:

- `ts_archived` + `getRememberedArchivedReportsThatShouldBeInvalidated` should be the most accurate solution
- we don't expect so many entries for archives to be invalidated, that the option table query can become a bottleneck (via `getRememberedArchivedReports...`, it probably has to run once for every `Archive::getDataTable` call)

If the option table query should not be a problem, I would switch the approach. Messing with the archiving can always be done later, once any information surfaces that warrants the risk/effort.

File: tests/PHPUnit/Integration/Archive/ArchiveStateMetadataTest.php [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+        self::assertSame(
+            $expectedFinalArchiveState,
+            $dataTable->getMetadata(DataTable::ARCHIVE_STATE_METADATA_NAME)
+        );
+    }
+
+    /**
+     * @return iterable<string, array>
+     */
+    public function archiveStateData(): iterable
```
Comment (michalkleiner) [inline tests/PHPUnit/Integration/Archive/ArchiveStateMetadataTest.php]: to more or less keep it in line with existing data provider methods
```suggestion
    public function getArchiveStateTestData(): iterable
```

File: tests/PHPUnit/Unit/Archive/ArchiveStateTest.php [hunk 10]
```diff
@@ -0,0 +1,304 @@
...
+            ->method('getDaysWithRememberedInvalidationsForSite')
+            ->willReturn($daysWithRememberedInvalidations);
+
+        $archiveState = new ArchiveState($archiveInvalidator);
+        $archiveState->addMetadataToResultCollection($collection, $archiveData, $archiveIds, $archiveStates);
+
+        $this->assertMetadataState($expectedMetadataState, $collection);
+    }
+
+    public function dataArchiveState(): iterable
```
Comment (michalkleiner) [inline tests/PHPUnit/Unit/Archive/ArchiveStateTest.php]:
```suggestion
    public function getArchiveStateTestData(): iterable
```

File: tests/PHPUnit/Unit/Archive/ArchiveStateTest.php [hunk 11]
```diff
@@ -0,0 +1,304 @@
...
+
+        $collection = $this->createCollection($archiveData);
+
+        $archiveState = new ArchiveState($archiveInvalidator);
+        $archiveState->addMetadataToResultCollection($collection, $archiveData, $archiveIds, $archiveStates);
+
+        $this->assertMetadataState($expectedState, $collection);
+    }
+
+    public function dataCheckTsArchivedWithBorderTimezones(): iterable
```
Comment (michalkleiner) [inline tests/PHPUnit/Unit/Archive/ArchiveStateTest.php]:
```suggestion
    public function getCheckTsArchivedWithBorderTimezonesTestData(): iterable
```

**Comment by michalkleiner** [comment]
Read the code and it looks ok to me, just a few minor suggestions around tests. Haven't done any local testing as I think there are others that understand archiving better.

**Review by caddoo — APPROVED** [review]

**Pull request merged**